### PR TITLE
feat: basic support OPA plugin

### PR DESCRIPTION
**Pull request by bzp2010**
### What this PR does / why we need it:

Add OPA plugin to support API access control using OPA services.
implement #5714 

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/plugins/opa.lua [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+local _M = {
+    version = 0.1,
+    priority = 2001,
+    name = "opa",
+    schema = schema,
+}
+
+
+function _M.check_schema(conf)
+    local ok, err = core.schema.check(schema, conf)
```
Comment (spacewander) [inline apisix/plugins/opa.lua:56]:
```suggestion
    return core.schema.check(schema, conf)
```

File: apisix/plugins/opa.lua [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    local res, err = httpc:request_uri(endpoint, params)
+
+    -- block by default when decision is unavailable
+    if not res or err then
+        core.log.error("failed to process OPA decision, err: ", err)
+        return 403
+    end
+
+    -- parse the results of the decision
+    local ret = core.json.decode(res.body).result
```
Comment (spacewander) [inline apisix/plugins/opa.lua:96]: Need to check if decode succeeded?

File: apisix/core/request.lua [hunk 3]
```diff
@@ -269,6 +269,15 @@ function _M.get_port(ctx)
 end
 
 
+function _M.get_path(ctx, original)
```
Comment (spacewander) [inline apisix/core/request.lua:272]: The `original` is unused?

File: apisix/plugins/opa/helper.lua [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+local core     = require("apisix.core")
+local ngx_var  = ngx.var
+local ngx_time = ngx.time
+
+local _M = {}
+
+
+local function build_var(conf, ctx)
+    return {
+        server_addr = ngx_var.server_addr,
```
Comment (spacewander) [inline apisix/plugins/opa/helper.lua:27]: Let's use ctx.var instead.

File: t/plugin/opa.t [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+property "host" is required
+
+
+
+=== TEST 3: wrong type for `host`
+--- config
+    location /t {
+        content_by_lua_block {
+            local plugin = require("apisix.plugins.opa")
+            local ok, err = plugin.check_schema({host = 3233, package = "example", decision = "allow"})
```
Comment (spacewander) [inline t/plugin/opa.t:78]: Can we use table drive test for check_schema?

**Review by anderseknert — COMMENTED** [review]
Nice! I can't comment on the Lua bits, but what I read looks sensible. Some random thoughts:

* I would probably just let the user configure the full OPA URL rather than separating host, package and decision. In some cases you'll actually query a whole package, which is fine for OPA.
* The policy example (and from what I understand, the Lua code) only deals with boolean decisions. Maybe good to keep it simple, but could also be useful to allow OPA to return for example a reason for denying a request, and surface that to the caller.
* I don't see much in terms of documentation. Will that be added in a later PR?

But all things considered, great work on this! 👏

**Comment by bzp2010** [comment]
Hi, @anderseknert. Glad to see your comments.

> I would probably just let the user configure the full OPA URL rather than separating host, package and decision. In some cases you'll actually query a whole package, which is fine for OPA.

Do you mean to merge host, package, decision together and use a single field (eg. `url`)?

> The policy example (and from what I understand, the Lua code) only deals with boolean decisions. Maybe good to keep it simple, but could also be useful to allow OPA to return for example a reason for denying a request, and surface that to the caller. I don't see much in terms of documentation. Will that be added in a later PR?

Features such as custom response codes, response headers and error causes will be added later in the PR. It will basically follow the roadmap section in #5714 to implement it.

**Comment by anderseknert** [comment]
> Do you mean to merge host, package, decision together and use a single field (eg. `url`)?

Yeah. Either way is fine though :)
 
> Features such as custom response codes, response headers and error causes will be added later in the PR. It will basically follow the roadmap section in #5714 to implement it.

Oh, nice! I had missed the roadmap section. Thanks for bringing that to my attention! Sounds like a good plan 👍

**Comment by bzp2010** [comment]
Ok, In that case, I would start by keeping the current way of combining URLs, which seems to provide better flexibility. XD

File: apisix/plugins/opa/helper.lua (hunk 4 above)
Comment (bzp2010) [reply apisix/plugins/opa/helper.lua:27]: This is a missing unmodified, previously I was using `reuqest_uri` var and separate `?` way to get pure paths, after checking the documentation I found that I can get pure paths using `uri` var. I will fix it.

File: t/plugin/opa.t (hunk 5 above)
Comment (bzp2010) [reply t/plugin/opa.t:78]: Yes, I think that's a good way, do we have any current examples of doing this for reference?

File: apisix/plugins/opa.lua (hunk 1 above)
Comment (bzp2010) [reply apisix/plugins/opa.lua]: fixed

File: apisix/plugins/opa.lua (hunk 2 above)
Comment (bzp2010) [reply apisix/plugins/opa.lua]: added

File: apisix/plugins/opa/helper.lua (hunk 4 above)
Comment (bzp2010) [reply apisix/plugins/opa/helper.lua]: fixed

File: apisix/core/request.lua (hunk 3 above)
Comment (bzp2010) [reply apisix/core/request.lua]: Yes, a useless code that I forgot to change.

Comment (bzp2010) [reply apisix/core/request.lua]: fixed

**Comment by anderseknert** [comment]
> Ok, In that case, I would keeping the current way of combining URLs, which seems to provide better flexibility. XD

I'm not sure that it's more flexible. It seems that currently all the fields are required to build the URL:

```lua
required = {"host", "package", "decision"}
```
But while not super common, it is perfectly legitimate to only query OPA for only a package, like:

```
http://localhost:8181/v1/data/policy
```

When OPA is queried on the package level, it will evaluate all the rules under that package and return the result. So a policy that looks like the below:

```rego
package policy

default allow = false

allow {
    input.user.roles[_] == "admin"
}

reason = "User needs to be admin" {
    not allow
}
```

When queried at `http://localhost:8181/v1/data/policy`, would respond with the result from evaluating both rules:

```json
{
    "allow": false,
    "reason": "User needs to be admin"
}
```

In this case, we don't query a rule (or decision as it's called in this code) directly, but rather all the rules in the package.

**Comment by bzp2010** [comment]
Oh, I see, I will change it later. I don't know OPA very well, thank you for pointing out my problem.

File: t/plugin/opa.t (hunk 5 above)
Comment (spacewander) [reply t/plugin/opa.t:78]: You can take this as an example: https://github.com/apache/apisix/blob/c178435d7ada4eeb713d9a1688fb5f54f971abdf/t/plugin/gzip.t#L494

File: apisix/plugins/opa.lua [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        core.log.error("failed to process OPA decision, err: ", err)
+        return 403
+    end
+
+    -- parse the results of the decision
+    local data, err = core.json.decode(res.body)
+
+    if err then
+        core.log.error("invalid response body: ", res.body, " err: ", err)
+        return 403
```
Comment (tokers) [inline apisix/plugins/opa.lua:95]: Should we return `500` when the json decoding is fail.y

Comment (bzp2010) [reply apisix/plugins/opa.lua]: changed

File: apisix/plugins/opa.lua [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        core.log.error("failed to process OPA decision, err: ", err)
+        return 403
+    end
+
+    -- parse the results of the decision
+    local data, err = core.json.decode(res.body)
+
+    if err then
+        core.log.error("invalid response body: ", res.body, " err: ", err)
+        return 500
```
Comment (spacewander) [inline apisix/plugins/opa.lua:94]:
```suggestion
        return 503
```
for known error

File: t/plugin/opa.t [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+property "policy" is required
+
+
+
+=== TEST 3: wrong type for `host`
+--- config
+    location /t {
+        content_by_lua_block {
+            local plugin = require("apisix.plugins.opa")
+            local ok, err = plugin.check_schema({host = 3233, policy = "example/allow"})
```
Comment (spacewander) [inline t/plugin/opa.t:78]: Don't forget https://github.com/apache/apisix/pull/5734#discussion_r766265874

File: t/plugin/opa.t (hunk 5 above)
Comment (bzp2010) [reply t/plugin/opa.t]: changed

File: apisix/plugins/opa.lua (hunk 7 above)
Comment (bzp2010) [reply apisix/plugins/opa.lua]: OK, changed

File: t/plugin/opa.t (hunk 8 above)
Comment (bzp2010) [reply t/plugin/opa.t]: I've switched to using TDT mode and waiting for CI to see how it turns out.

Comment (bzp2010) [reply t/plugin/opa.t]: TDT changed

**Review by spacewander — APPROVED** [review]

**Comment by bzp2010** [comment]
ping @tokers @tzssangglass @shuaijinchao PTAL, thanks

**Review by shuaijinchao — APPROVED** [review]

**Review by tzssangglass — APPROVED** [review]

**Comment by moonming** [comment]
@bzp2010 please add OPA into the README, thanks.

**Pull request merged**